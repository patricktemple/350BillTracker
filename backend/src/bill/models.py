import enum
from uuid import uuid4

import flask
from sqlalchemy import Boolean, Column, Enum, ForeignKey, Integer, Text, sql
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import foreign, relationship, remote

from ..models import TIMESTAMP, UUID, db
from ..user.models import User
from ..utils import now

DEFAULT_TWITTER_SEARCH_TERMS = [
    "solar",
    "climate",
    "wind power",
    "renewable",
    "fossil fuel",
]


class StateChamber(enum.Enum):
    SENATE = 1
    ASSEMBLY = 2


class Bill(db.Model):
    """
    Base table for all bills, both city and state. Contains any info that's
    relevant for both kinds of bills.

    Bills are polymorphic. This table has a "type" which specifies whether it's
    a city or a state bill, and there will be an associated row in the CityBill
    and StateBill with more info on it."""

    __tablename__ = "bills"

    class BillType(enum.Enum):
        CITY = 1
        STATE = 2

    id = Column(UUID, primary_key=True, default=uuid4)

    # For city this is the "MatterName", for state it is the "title"
    name = Column(Text, nullable=False)

    # For city this is the "title", for state this is the "summary"
    description = Column(Text, nullable=False)

    # Info on child objects:
    type = Column(Enum(BillType), nullable=False)

    # If this is is a city bill, points to more specific data on that.
    city_bill = relationship(
        "CityBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # If this is is a state bill, points to more specific data on that.
    state_bill = relationship(
        "StateBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # Data we track
    notes = Column(Text, nullable=False, server_default="")
    nickname = Column(Text, nullable=False, server_default="")
    twitter_search_terms = Column(
        ARRAY(Text), nullable=False, default=DEFAULT_TWITTER_SEARCH_TERMS
    )

    power_hours = relationship(
        "PowerHour", back_populates="bill", cascade="all, delete"
    )
    attachments = relationship(
        "BillAttachment", back_populates="bill", cascade="all, delete"
    )

    @property
    def display_name(self):
        return self.nickname if self.nickname else self.name

    @property
    def tracked(self):
        return True

    @property
    def status(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.status

        senate_status = (
            self.state_bill.senate_bill.status
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_status = (
            self.state_bill.assembly_bill.status
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_status} / {assembly_status}"

    @property
    def code_name(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.file

        senate_print_no = (
            self.state_bill.senate_bill.base_print_no
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_print_no = (
            self.state_bill.assembly_bill.base_print_no
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_print_no} / {assembly_print_no} from {self.state_bill.session_year} session"
    
    user_bill_settings = relationship("UserBillSettings", back_populates="bill", cascade="all, delete-orphan")


class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")


class CityBill(db.Model):
    """
    City-specific details about an NYC bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name).
    """

    __tablename__ = "city_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)

    # ID in the City Council API
    city_bill_id = Column(Integer, nullable=False, unique=True, index=True)

    # Code name often used to discuss the bill, like Int 2317-2021
    file = Column(Text, nullable=False)

    # The parent Bill object that represents this.
    bill = relationship("Bill", back_populates="city_bill", lazy="joined")

    intro_date = Column(TIMESTAMP, nullable=False)

    status = Column(Text, nullable=False)

    # Which amendment version is currently active. We only track one at a time,
    # and sponsorships are tied to that particular version.
    active_version = Column(Text, nullable=False)

    sponsorships = relationship(
        "CitySponsorship", back_populates="city_bill", cascade="all, delete"
    )

    # Committee name
    council_body = Column(Text)


# TODO: Understand bill substitution and change if needed
class StateBill(db.Model):
    """
    State-specific details about a state bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name)."""

    __tablename__ = "state_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    bill = relationship(Bill, back_populates="state_bill", lazy="joined")

    # The start of the 2-year legislative session this belongs to.
    session_year = Column(Integer, nullable=False)

    senate_bill = relationship(
        "SenateBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )
    assembly_bill = relationship(
        "AssemblyBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )


class StateChamberMixin:
    """
    Mixin for a bill in either the Senate or Assembly chamber. The have
    identical structure but are kept in separate tables."""

    active_version = Column(Text, nullable=False)
    status = Column(Text, nullable=False)
    base_print_no = Column(Text, nullable=False)

    @declared_attr
    def bill_id(self):
        return Column(UUID, ForeignKey(StateBill.bill_id), primary_key=True)

    # The Senate and the Assembly each run separate websites, and each website can
    # lookup both senate and assembly bills. So they're redundant websites with
    # very different UI. Therefore senate and assembly bills each have a
    # both assembly and senate websites.
    @property
    def senate_website(self):
        return f"https://www.nysenate.gov/legislation/bills/{self.state_bill.session_year}/{self.base_print_no}"

    @property
    def assembly_website(self):
        return f"https://nyassembly.gov/leg/?term={self.state_bill.session_year}&bn={self.base_print_no}"


class SenateBill(db.Model, StateChamberMixin):
    __tablename__ = "senate_bills"

    state_bill = relationship(StateBill, back_populates="senate_bill")
    sponsorships = relationship(
        "SenateSponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class AssemblyBill(db.Model, StateChamberMixin):
    __tablename__ = "assembly_bills"

    state_bill = relationship(StateBill, back_populates="assembly_bill")
    sponsorships = relationship(
        "AssemblySponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class UserBillSettings(db.Model):
    __tablename__ = "user_bill_settings"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    user_id = Column(UUID, ForeignKey(User.id), primary_key=True)

    send_bill_update_notifications = Column(
        Boolean, nullable=False, server_default=sql.false()
    )

    user = relationship(User, lazy="joined", back_populates="bill_settings")
    bill = relationship(Bill, back_populates="user_bill_settings")


Bill.viewer_settings = relationship(
    UserBillSettings,
    primaryjoin=(remote(UserBillSettings.bill_id) == foreign(Bill.id))
    & (
        UserBillSettings.user_id
        == sql.bindparam(
            "request_user_id", callable_=lambda: flask.g.request_user_id
        )
    ),
    viewonly=True,
)
