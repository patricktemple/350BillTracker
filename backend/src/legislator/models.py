from uuid import uuid4

from sqlalchemy import Column, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db


class Legislator(db.Model):
    __tablename__ = "legislators"

    # These come from the API
    id = Column(Integer, primary_key=True)
    name = Column(Text, nullable=False)
    term_start = Column(TIMESTAMP)
    term_end = Column(TIMESTAMP)
    email = Column(Text)
    district_phone = Column(Text)
    legislative_phone = Column(Text)

    sponsorships = relationship("BillSponsorship", back_populates="legislator")
    staffers = relationship("Staffer", back_populates="legislator")

    borough = Column(Text)
    website = Column(Text)

    # These are added by our static data
    twitter = Column(Text)  # exclude the @ symbol
    party = Column(Text)

    # Track our own info on the bill.
    notes = Column(Text)

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def twitter_url(self):
        return (
            f"https://www.twitter.com/{self.twitter}" if self.twitter else None
        )


class Staffer(db.Model):
    __tablename__ = "staffers"

    id = Column(UUID, primary_key=True, default=uuid4)
    name = Column(Text, nullable=False)
    title = Column(Text)
    email = Column(Text)
    phone = Column(Text)
    legislator_id = Column(
        Integer, ForeignKey("legislators.id"), nullable=False
    )
    twitter = Column(Text)

    legislator = relationship("Legislator", back_populates="staffers")

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def display_string(self):
        contact_methods = [self.phone, self.email, self.display_twitter]
        contact_methods = [c for c in contact_methods if c]
        contact_string = ", ".join(contact_methods)
        if not contact_string:
            contact_string = "No contact info"
        title_string = f"{self.title} - " if self.title else ""

        return f"{title_string}{self.name} ({contact_string})"
