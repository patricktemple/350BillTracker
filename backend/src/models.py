from uuid import uuid4

from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Integer,
    Text,
    TypeDecorator,
    sql,
)
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.dialects.postgresql import TIMESTAMP as _TIMESTAMP
from sqlalchemy.dialects.postgresql import UUID as _UUID
from sqlalchemy.orm import relationship

from .app import app
from .utils import now

db = SQLAlchemy(app)


class UUID(TypeDecorator):
    impl = _UUID(as_uuid=True)


class TIMESTAMP(TypeDecorator):
    impl = _TIMESTAMP(timezone=True)


# TODO: UUIDs for some PKs?
class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(Integer, primary_key=True)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")
