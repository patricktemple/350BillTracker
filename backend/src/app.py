import logging

from flask import Flask
from flask_marshmallow import Marshmallow

from .settings import DATABASE_URL

app = Flask(__name__)
app.config["SQLALCHEMY_DATABASE_URI"] = DATABASE_URL
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False

marshmallow = Marshmallow(app)

logging.basicConfig(
    format="%(asctime)s [%(levelname)s] %(message)s",
    level=logging.INFO,
    datefmt="%Y-%m-%d %H:%M:%S",
)

# TODO: Can simplify this by just importing the views and letting the rest happen
from . import models, cron, views  # noqa: F401 isort:skip
from .bill import models as bill_models  # noqa: F401 isort:skip
from .bill import views as bill_views
from .person import views as person_views
from .sponsorship import views as sponsorship_views
from .user import views as user_views

from .person import models as person_models  # noqa: F401 isort:skip
from .sponsorship import models as sponsorship_models  # noqa: F401 isort:skip
from .user import models as user_models  # noqa: F401 isort:skip
