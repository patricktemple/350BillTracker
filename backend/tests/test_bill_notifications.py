from unittest.mock import patch
from uuid import uuid4

import responses

from src.app import app
from src.bill.models import AssemblyBill, Bill, CityBill, SenateBill, StateBill
from src.bill_notifications import (
    BillDiff,
    BillSnapshot,
    _calculate_bill_diffs,
    _convert_bill_diff_to_template_variables,
    _get_bill_update_subject_line,
    send_bill_update_notifications,
)
from src.models import db
from src.person.models import CouncilMember, Person
from src.sponsorship.models import CitySponsorship
from src.user.models import User
from src.utils import now


# TODO: Switch to Factory
def add_test_bill(city_bill_id, status) -> Bill:
    bill = Bill(
        id=uuid4(),
        name=f"{city_bill_id} name",
        description="description",
        type=Bill.BillType.CITY,
    )
    bill.city_bill = CityBill(
        city_bill_id=city_bill_id,
        file=f"{city_bill_id} file",
        status=status,
        intro_date=now(),
        active_version="A",
    )
    db.session.add(bill)
    return bill


def add_test_council_member(city_council_person_id) -> Person:
    person = Person(
        id=uuid4(),
        name=f"{city_council_person_id} name",
        type=Person.PersonType.COUNCIL_MEMBER,
    )
    person.council_member = CouncilMember(
        city_council_person_id=city_council_person_id
    )
    db.session.add(person)
    return person


def add_test_sponsorship(*, bill, person):
    sequence = 0

    def impl():
        nonlocal sequence
        sponsorship = CitySponsorship(
            bill_id=bill.id,
            council_member_id=person.id,
            sponsor_sequence=sequence,
        )
        db.session.add(sponsorship)
        sequence += 1
        return sponsorship

    return impl()


def test_calculate_bill_diffs():
    bill_1 = add_test_bill(1, "Enacted")
    bill_2 = add_test_bill(2, "Enacted")
    council_member_1 = add_test_council_member(1)
    council_member_2 = add_test_council_member(2)
    council_member_3 = add_test_council_member(3)
    add_test_sponsorship(bill=bill_1, person=council_member_1)
    add_test_sponsorship(bill=bill_1, person=council_member_3)
    add_test_sponsorship(bill=bill_2, person=council_member_3)

    db.session.commit()

    snapshots = {
        # This should change status, lose sponsor 2 and gain sponsor 3
        bill_1.id: BillSnapshot(
            "Committee", [council_member_1.id, council_member_2.id]
        ),
        # This should be unchanged
        bill_2.id: BillSnapshot("Enacted", [council_member_3.id]),
    }
    diffs = _calculate_bill_diffs(snapshots)

    assert len(diffs) == 1
    diff = diffs[0]

    assert diff.added_sponsor_names == [council_member_3.name]
    assert diff.removed_sponsor_names == [council_member_2.name]
    assert diff.old_status == "Committee"
    assert diff.city_bill.status == "Enacted"


def test_email_contents__status_changed():
    bill = add_test_bill(1, "Enacted")
    diff = BillDiff(old_status="Committee", city_bill=bill.city_bill)

    subject = _get_bill_update_subject_line([diff])
    assert subject == f"{bill.city_bill.file}'s status changed to Enacted"

    variables = _convert_bill_diff_to_template_variables(diff)
    assert variables == {
        "file": bill.city_bill.file,
        "display_name": bill.display_name,
        "status_text": "Status: Committee --> Enacted",
        "status_color": "blue",
        "sponsor_text": "0 sponsors (unchanged)",
        "sponsor_color": "black",
    }


def test_email_contents__sponsors_added():
    bill = add_test_bill(1, "Enacted")
    diff = BillDiff(
        old_status="Enacted",
        city_bill=bill.city_bill,
        added_sponsor_names=["Brad Lander", "Jamaal Bowman"],
    )

    # It uses the current sponsor count to figure out the previous, so
    # add some sponsors in the DB so it doesn't report a negative number
    for i in range(3):
        person = add_test_council_member(i)
        add_test_sponsorship(bill=bill, person=person)

    db.session.commit()

    subject = _get_bill_update_subject_line([diff])
    assert subject == f"{bill.city_bill.file} gained 2 sponsors"

    variables = _convert_bill_diff_to_template_variables(diff)
    assert variables == {
        "file": bill.city_bill.file,
        "display_name": bill.display_name,
        "status_text": "Status: Enacted (unchanged)",
        "status_color": "black",
        "sponsor_text": "1 sponsors --> 3 sponsors (gained Brad Lander, Jamaal Bowman)",
        "sponsor_color": "blue",
    }


def test_email_subject__1_sponsor_added():
    bill = add_test_bill(1, "Enacted")
    diff = BillDiff(
        old_status="Enacted",
        city_bill=bill.city_bill,
        added_sponsor_names=["Brad Lander"],
    )

    for i in range(3):
        person = add_test_council_member(i)
        add_test_sponsorship(bill=bill, person=person)

    db.session.commit()

    subject = _get_bill_update_subject_line([diff])
    assert subject == f"{bill.city_bill.file} gained sponsor Brad Lander"


def test_email_contents__sponsor_removed():
    bill = add_test_bill(1, "Enacted")
    diff = BillDiff(
        old_status="Enacted",
        city_bill=bill.city_bill,
        removed_sponsor_names=["Brad Lander"],
    )

    subject = _get_bill_update_subject_line([diff])
    assert subject == f"{bill.city_bill.file} lost sponsor Brad Lander"

    variables = _convert_bill_diff_to_template_variables(diff)
    assert variables == {
        "file": bill.city_bill.file,
        "display_name": bill.display_name,
        "status_text": "Status: Enacted (unchanged)",
        "status_color": "black",
        "sponsor_text": "1 sponsors --> 0 sponsors (lost Brad Lander)",
        "sponsor_color": "blue",
    }


def test_email_contents__sponsor_added_and_removed():
    bill = add_test_bill(1, "Enacted")
    diff = BillDiff(
        old_status="Enacted",
        city_bill=bill.city_bill,
        added_sponsor_names=["Jamaal Bowman"],
        removed_sponsor_names=["Brad Lander"],
    )

    person = add_test_council_member(1)
    add_test_sponsorship(bill=bill, person=person)

    db.session.commit()

    subject = _get_bill_update_subject_line([diff])
    assert subject == f"{bill.city_bill.file} gained and lost sponsors"

    variables = _convert_bill_diff_to_template_variables(diff)
    assert variables == {
        "file": bill.city_bill.file,
        "display_name": bill.display_name,
        "status_text": "Status: Enacted (unchanged)",
        "status_color": "black",
        "sponsor_text": "1 sponsors --> 1 sponsors (gained Jamaal Bowman, lost Brad Lander)",
        "sponsor_color": "blue",
    }


@responses.activate
@patch("src.ses.client")
def test_state_bill_ignored(mock_ses_client):
    bill = Bill(
        id=uuid4(),
        name="state bill",
        description="description",
        type=Bill.BillType.STATE,
    )
    bill.state_bill = StateBill(
        session_year=2021,
    )
    bill.state_bill.senate_bill = SenateBill(
        base_print_no="S1234", active_version="A", status="Signed by governor"
    )
    bill.state_bill.assembly_bill = AssemblyBill(
        base_print_no="A1234", active_version="", status="Signed by governor"
    )
    db.session.add(bill)
    db.session.commit()

    snapshots = {}

    with app.app_context():
        send_bill_update_notifications(snapshots)

    mock_ses_client.send_email.assert_not_called()


@responses.activate
@patch("src.ses.client")
def test_send_email_notification_end_to_end(mock_ses_client):
    user_to_notify = User(
        id=uuid4(),
        name="User to notify",
        email="user@example.com",
        send_bill_update_notifications=True,
    )
    db.session.add(user_to_notify)
    other_user = User(
        id=uuid4(), name="Someone else", email="wrong@example.com"
    )
    db.session.add(other_user)

    bill = add_test_bill(1, "Enacted")

    db.session.commit()

    snapshots = {
        bill.id: BillSnapshot("Committee", sponsor_ids=[]),
    }

    def side_effect(*, Destination, Message, Source):
        assert Destination["ToAddresses"] == ["user@example.com"]
        assert "Enacted" in Message["Body"]["Html"]["Data"]
        assert "Enacted" in Message["Body"]["Text"]["Data"]
        assert "Enacted" in Message["Subject"]["Data"]
        return {"MessageId": 1}

    mock_ses_client.send_email.side_effect = side_effect

    with app.app_context():
        send_bill_update_notifications(snapshots)

    mock_ses_client.send_email.assert_called_once()
