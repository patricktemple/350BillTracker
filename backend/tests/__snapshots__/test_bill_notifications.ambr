# name: test_email_contents__city_sponsor_added_and_removed
  '
  
  SUBJECT--------------------------------------------
  City bill Intro 2317 gained and lost sponsors
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">Intro 2317 - Gass Free NYC</div>
        <div style="color: black">Status: Enacted (unchanged)</div>
        <div style="color: blue">1 sponsors --&gt; 1 sponsors (gained Jamaal Bowman, lost Brad Lander)</div>
      </div>
      
  
      
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  Intro 2317 - Gass Free NYC
  Status: Enacted (unchanged)
  1 sponsors --> 1 sponsors (gained Jamaal Bowman, lost Brad Lander)
  
  
  
  
  '
---
# name: test_email_contents__city_sponsor_removed
  '
  
  SUBJECT--------------------------------------------
  City bill Intro 2317 lost sponsor Brad Lander
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">Intro 2317 - Gas Free NYC</div>
        <div style="color: black">Status: Enacted (unchanged)</div>
        <div style="color: blue">1 sponsors --&gt; 0 sponsors (lost Brad Lander)</div>
      </div>
      
  
      
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  Intro 2317 - Gas Free NYC
  Status: Enacted (unchanged)
  1 sponsors --> 0 sponsors (lost Brad Lander)
  
  
  
  
  '
---
# name: test_email_contents__city_sponsors_added
  '
  
  SUBJECT--------------------------------------------
  City bill Intro 2317 gained 2 sponsors
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">Intro 2317 - Gass Free NYC</div>
        <div style="color: black">Status: Enacted (unchanged)</div>
        <div style="color: blue">1 sponsors --&gt; 3 sponsors (gained Brad Lander, Jamaal Bowman)</div>
      </div>
      
  
      
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  Intro 2317 - Gass Free NYC
  Status: Enacted (unchanged)
  1 sponsors --> 3 sponsors (gained Brad Lander, Jamaal Bowman)
  
  
  
  
  '
---
# name: test_email_contents__city_status_changed
  '
  
  SUBJECT--------------------------------------------
  City bill Intro 2317's status changed to Enacted
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">Intro 2317 - Gas Free NYC</div>
        <div style="color: blue">Status: Committee --&gt; Enacted</div>
        <div style="color: black">0 sponsors (unchanged)</div>
      </div>
      
  
      
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  Intro 2317 - Gas Free NYC
  Status: Committee --> Enacted
  0 sponsors (unchanged)
  
  
  
  
  '
---
# name: test_email_contents__state_both_status_changed
  '
  
  SUBJECT--------------------------------------------
  State bill A4264/S4264 was updated in both chambers
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      
  
      
      <div style="font-size: 1.5rem; font-weight: bold">State bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">CCIA</div>
        
          <div style="font-weight: bold">S4264 - Senate</div>
          <div style="color: blue">Status: Committee --&gt; Signed by governor</div>
          <div style="color: black">0 sponsors (unchanged)</div>
        
          <div style="font-weight: bold">A4264 - Assembly</div>
          <div style="color: blue">Status: Committee --&gt; Signed by governor</div>
          <div style="color: black">0 sponsors (unchanged)</div>
        
      </div>
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  
  
  STATE BILLS
  
  CCIA
  
  S4264 - Senate
  Status: Committee --> Signed by governor
  0 sponsors (unchanged)
  
  A4264 - Assembly
  Status: Committee --> Signed by governor
  0 sponsors (unchanged)
  
  
  
  '
---
# name: test_email_contents__state_gained_senate_sponsor
  '
  
  SUBJECT--------------------------------------------
  State bill S4264 gained sponsor Jabari Brisport
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      
  
      
      <div style="font-size: 1.5rem; font-weight: bold">State bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">CCIA</div>
        
          <div style="font-weight: bold">S4264 - Senate</div>
          <div style="color: black">Status: Committee (unchanged)</div>
          <div style="color: blue">0 sponsors --&gt; 1 sponsors (gained Jabari Brisport)</div>
        
      </div>
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  
  
  STATE BILLS
  
  CCIA
  
  S4264 - Senate
  Status: Committee (unchanged)
  0 sponsors --> 1 sponsors (gained Jabari Brisport)
  
  
  
  '
---
# name: test_email_contents__state_lost_assembly_sponsor
  '
  
  SUBJECT--------------------------------------------
  State bill A234 lost sponsor Emily Gallagher
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      
  
      
      <div style="font-size: 1.5rem; font-weight: bold">State bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">CCIA</div>
        
          <div style="font-weight: bold">A234 - Assembly</div>
          <div style="color: black">Status: Committee (unchanged)</div>
          <div style="color: blue">1 sponsors --&gt; 0 sponsors (lost Emily Gallagher)</div>
        
      </div>
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  
  
  STATE BILLS
  
  CCIA
  
  A234 - Assembly
  Status: Committee (unchanged)
  1 sponsors --> 0 sponsors (lost Emily Gallagher)
  
  
  
  '
---
# name: test_email_contents__state_status_changed
  '
  
  SUBJECT--------------------------------------------
  State bill S4264's status changed to Signed by governor
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      
  
      
      <div style="font-size: 1.5rem; font-weight: bold">State bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">CCIA</div>
        
          <div style="font-weight: bold">S4264 - Senate</div>
          <div style="color: blue">Status: Committee --&gt; Signed by governor</div>
          <div style="color: black">0 sponsors (unchanged)</div>
        
      </div>
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  
  
  STATE BILLS
  
  CCIA
  
  S4264 - Senate
  Status: Committee --> Signed by governor
  0 sponsors (unchanged)
  
  
  
  '
---
# name: test_email_subject__1_city_sponsor_added
  '
  
  SUBJECT--------------------------------------------
  City bill Intro 2317 gained sponsor Brad Lander
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">Intro 2317 - Gas Free NYC</div>
        <div style="color: black">Status: Enacted (unchanged)</div>
        <div style="color: blue">2 sponsors --&gt; 3 sponsors (gained Brad Lander)</div>
      </div>
      
  
      
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  Intro 2317 - Gas Free NYC
  Status: Enacted (unchanged)
  2 sponsors --> 3 sponsors (gained Brad Lander)
  
  
  
  
  '
---
# name: test_send_email_notification_end_to_end
  '
  
  SUBJECT--------------------------------------------
  2 bills were updated
  
  BODY HTML------------------------------------------
  <html>
    <head>
    </head>
    <body>
      <p>The following bills were updated by the City Council.</p>
      
      <div style="font-size: 1.5rem; font-weight: bold">City bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">file - nickname</div>
        <div style="color: blue">Status: Not introduced --&gt; Enacted</div>
        <div style="color: black">0 sponsors (unchanged)</div>
      </div>
      
  
      
      <div style="font-size: 1.5rem; font-weight: bold">State bills</div>
      
      
      <div style="padding-bottom: 1rem">
        <div style="font-size: 1rem; font-weight: bold">nickname</div>
        
          <div style="font-weight: bold">S1234 - Senate</div>
          <div style="color: blue">Status: Introduced in senate --&gt; Committee</div>
          <div style="color: black">0 sponsors (unchanged)</div>
        
          <div style="font-weight: bold">A1234 - Assembly</div>
          <div style="color: blue">Status: Introduced in assembly --&gt; Voted</div>
          <div style="color: black">0 sponsors (unchanged)</div>
        
      </div>
      
    </body>
  </html>
  
  BODY TEXT-------------------------------------------
  The following bills were updated by the City Council.
  
  
  CITY BILLS
  
  file - nickname
  Status: Not introduced --> Enacted
  0 sponsors (unchanged)
  
  
  
  
  STATE BILLS
  
  nickname
  
  S1234 - Senate
  Status: Introduced in senate --> Committee
  0 sponsors (unchanged)
  
  A1234 - Assembly
  Status: Introduced in assembly --> Voted
  0 sponsors (unchanged)
  
  
  
  '
---
